src/zcash/address/mnemonic.cpp

```diff
@@ -9,30 +9,60 @@
+    auto seed = MnemonicSeed::ForPhrase(language, mnemonic).value();
+
+    // Verify that the seed data is valid entropy for unified spending keys at
+    // account 0 and at both the public & private chain levels for account 0x7FFFFFFF.
+    // It is not necessary to check for a valid diversified Sapling address at
+    // account 0x7FFFFFFF because derivation via the legacy path can simply search
+    // for a valid diversifier; unlike in the unified spending key case, diversifier
+    // indices don't need to line up with anything.
+    if (ZcashdUnifiedSpendingKey::ForAccount(seed, bip44CoinType, 0).has_value() &&
+        transparent::AccountKey::ForAccount(seed, bip44CoinType, ZCASH_LEGACY_ACCOUNT).has_value())  {
+        return seed;
+    } else {
+        return std::nullopt;
+    }
+}
+
 MnemonicSeed MnemonicSeed::Random(uint32_t bip44CoinType, Language language, size_t entropyLen)
 {
     assert(entropyLen >= 32);
     while (true) { // loop until we find usable entropy
         RawHDSeed entropy(entropyLen, 0);
         GetRandBytes(entropy.data(), entropyLen);
-        const char* phrase = zip339_entropy_to_phrase(language, entropy.data(), entropyLen);
-        SecureString mnemonic(phrase);
-        zip339_free_phrase(phrase);
-
-        // The phrase returned from zip339_entropy_to_phrase should always be a
-        // valid UTF-8 string; this `.value()` unwrap will correctly throw a
-        // `std::bad_optional_access` exception if that invariant does not hold.
-        auto seed = MnemonicSeed::ForPhrase(language, mnemonic).value();
-
-        // Verify that the seed data is valid entropy for unified spending keys at
-        // account 0 and at both the public & private chain levels for account 0x7FFFFFFF.
-        // It is not necessary to check for a valid diversified Sapling address at
-        // account 0x7FFFFFFF because derivation via the legacy path can simply search
-        // for a valid diversifier; unlike in the unified spending key case, diversifier
-        // indices don't need to line up with anything.
-        if (ZcashdUnifiedSpendingKey::ForAccount(seed, bip44CoinType, 0).has_value() &&
-            transparent::AccountKey::ForAccount(seed, bip44CoinType, ZCASH_LEGACY_ACCOUNT).has_value())  {
-            return seed;
+
+        auto seed = MnemonicSeed::FromEntropy(entropy, bip44CoinType, language);
+        if (seed.has_value()) {
+            return seed.value();
+        }
+    }
+}
+
+MnemonicSeed MnemonicSeed::FromLegacySeed(const HDSeed& legacySeed, uint32_t bip44CoinType, Language language)
+{
+    auto rawSeed = legacySeed.RawSeed();
+    assert(rawSeed.size() == 32);
+
+    unsigned char nonce = 0;
+    while(true) {
+        auto seed = MnemonicSeed::FromEntropy(rawSeed, bip44CoinType, language);
+        if (seed.has_value()) {
+            return seed.value();
+        } else {
+            assert(nonce < 255);
+            nonce++;
+            rawSeed[0] += nonce;
```

Comment:
This is adding 1, then 2, then 3..., which is not correct. It should just be `rawSeed[0]++;`.